src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -966,15 +970,37 @@ QString QgsSymbolLegendNode::evaluateLabel( const QgsExpressionContext &context,
 
 QgsExpressionContextScope *QgsSymbolLegendNode::createSymbolScope() const
 {
-  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
-
   QgsExpressionContextScope *scope = new QgsExpressionContextScope( tr( "Symbol scope" ) );
   scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), symbolLabel().remove( "[%" ).remove( "%]" ), true ) );
   scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
-  if ( vl )
+
+  if ( mLayerNode )
   {
-    scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_count" ), QVariant::fromValue( vl->featureCount( mItem.ruleKey() ) ), true ) );
+    QString symbolExp;
+    QMap<QString, QString> modelstyles = model()->layerStyleOverrides();
+    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+    if ( vl )
+    {
+      QgsMapLayerStyleOverride styleOverride( vl );
+      if ( modelstyles.contains( vl->id() ) )
+        styleOverride.setOverrideStyle( modelstyles.value( vl->id() ) );
```

Comment:
Applying an override is a very slow operation -- is this actually required? Or just copied/pasted from elsewhere?

Reply:
It's required AFAIK, previously the implementation called the renderer directly, but if the style of a given layer in the canvas was different than in the layout ( cause by using a theme in the layout) then the expression could not be retreived by the renderer as `legendKeyToExpression` could not return the value.

If you think there is another way to retraive the expression from the renderer without using the override, this would be good. Though maybe some logic could be implemented to bypass the override and only perform the operation when required.

I'll let you make the call on what could be improved and I'll rework the PR as needed.

Reply:
I'd have expected that this was already handled prior to calling any of this code -- otherwise even the simple case of a map with a different style having different color fills would be broken. 

Can you try removing this code and retest?

Reply:
As suggested, I removed the style override and both with local manual tests and in the test image, the lines are just NULL.